docs/introduction.md

```diff
@@ -0,0 +1,139 @@
+
+## What's a `Validator`?
+
+A `Validator` uses JSON Schema to ensure that when you try to run your job,
+you have the correct arguments. For example, in the example above the
+`Validator` is the empty json object `{}`. This means that providing any arguments
+will fail. That model is not a very useful model. We could instead require a green
+band, a red band, and a GeoTIFF location with the following JSON schema:
+
+```json
+{
+  "definitions": {},
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "http://example.com/root.json",
+  "type": "object",
+  "title": "The Root Schema",
+  "required": [
+    "greenBand",
+    "redBand",
+    "tiffLocation"
+  ],
+  "properties": {
+    "greenBand": {
+      "$id": "#/properties/greenBand",
+      "type": "integer",
+      "title": "The Greenband Schema",
+      "default": 0,
+      "examples": [
+        3
+      ]
+    },
+    "redBand": {
+      "$id": "#/properties/redBand",
+      "type": "integer",
+      "title": "The Redband Schema",
+      "default": 0,
+      "examples": [
+        2
+      ]
+    },
+    "tiffLocation": {
+      "$id": "#/properties/tiffLocation",
+      "type": "string",
+      "title": "The Tifflocation Schema",
+      "default": "",
+      "examples": [
+        "s3://cool-bucket/image.tiff"
+      ],
+      "format": "uri"
+    }
+  }
+}
+```
+
+With this JSON schema attached to our model, it's impossible to submit
+improperly formatted arguments.
+
+Over time, your model's input requirements may change. In that case it makes
+sense to make a new model with a different validator. Since writing JSON schema
+by hand is tedious, it's helpful to use [`json-schema.net`](https://jsonschema.net/)
+to generate schemas from examples, then edit those to match your needs.
+
+## What's a `Prediction`?
+
+A `Prediction` is a single run of a model with specific inputs. Predictions are
+created when you `POST` a model id and arguments to `/api/predictions`.
+
+```scala mdoc
+val modelId = UUID.randomUUID
+
+val prediction = Prediction.Create(
+  modelId, // the modelId to associate with this prediction
+  ().asJson // the arguments to pass to the model
+)
+
+// What you'll POST to the API to create a prediction
+prediction.asJson.spaces2
+```
+
+The model's JSON schema is used to validate the prediction's arguments.
+If validation passes, Granary will insert a record for this prediction and submit
+a job to AWS Batch with the resources configured on the model. At this stage,
+it will also create a single-use webhook for updating the prediction. This webhook
+will be at `/api/predictions/{predictionId}/results/{webhookId}` and accepts two
+kinds of messages:
+
+```scala mdoc
+// JSON of what you should send to the webhook if the prediction failed
+PredictionFailure("everything went wrong").asJson.spaces2
+
+// JSON of what you should send to the webhook if the prediction succeeded
+PredictionSuccess("s3://where/the/results/live.json").asJson.spaces2
+```
```

Comment:
To clarify my understanding, is it expected that the the AWS Batch job will POST to one of these webhooks to signal job status. It may make things more clear if the "you" in "JSON of what you should send to the webhook" is changed to the specific service within the architecture that makes the POST. I am assuming it is not the creator of the `Prediction`.

If my assumptions are correct, Batch Job is responsible for calling the webhooks, how are abrupt crashes handled (out of memory, for example)?

Reply:
> Batch Job is responsible for calling the webhooks, how are abrupt crashes handled?

Short answer is they aren't. Error-handling within the batch job is the user's responsibility. We've chosen to be extremely hands-off with the ways that batch jobs can fail, since that space is basically infinite.

> It may make things more clear if the "you" in "JSON of what you should send to the webhook" is changed to the specific service within the architecture that makes the POST. I am assuming it is not the creator of the Prediction.

Correct, best-case scenario is that whatever is running inside the container makes the status update request when it's done. How that happens is pretty flexible -- Lambda could listen for completed batch jobs, the jobs themselves could do the update, it could be a developer's job to monitor and POST -- I'll clarify the case we think is the best one.

Reply:
I [attempted to clarify](https://github.com/raster-foundry/granary/pull/143/files#diff-0970123c5329762d23736c706f9c6caaR171)